### PR TITLE
Cast overflow varchar to decimal

### PR DESCRIPTION
**Pull request by LindsayWray**
Addresses issue #4978 nr. 35: Cast overflow. Example:
```SELECT '-3.4028234663852886e+38'::DECIMAL```

When casting number represented as varchar to a decimal value it would overflow when too large to fit into the underlying integer types of a decimal, now it returns a casting error instead.  
<strike>  I’ve handled this issue by adding numeric limits check which checks limits for int16, int32, int64, and huge-int required a specialisation check. </strike>
I just found out this functionality was already available in the codebase ! So used that instead

### CODE REVIEW

File: src/common/operator/cast_operators.cpp [hunk 1]
```diff
@@ -1601,8 +1602,14 @@ struct DecimalCastOperation {
 		}
 		state.digit_count++;
 		if (NEGATIVE) {
+			if (state.result < (NumericLimits<typename T::type_t>::Minimum() / 10)) {
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:1605]: nit: this could just be `NumericLimits<T>`, and we could remove the `type_t`, no?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good to me. One minor comment:

File: src/common/operator/cast_operators.cpp (hunk 1 above)
Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1605]: Ah there are different T's present here, within the scope of the handledigit function, T is the type of state (DecimalCastData) not the numeric type

Comment (Mytherin) [reply src/common/operator/cast_operators.cpp:1605]: Oh, I see, that is a bit confusing indeed. Perhaps instead rename one of the T's to remove that ambiguity?

Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1605]: I've altered it!

File: src/common/operator/cast_operators.cpp [hunk 2]
```diff
@@ -1601,8 +1602,14 @@ struct DecimalCastOperation {
...
+				return false;
+			}
 			state.result = state.result * 10 - digit;
 		} else {
+			if (state.result > (NumericLimits<typename T::type_t>::Maximum() / 10)) {
```
Comment (Mytherin) [inline src/common/operator/cast_operators.cpp:1610]: Now that the template parameter is renamed we should be able to get rid of the `type_t`, right? :)

Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1610]: T is here the state, which is a DecimalCastData struct, not a numeric type. And i want the limit of the type with which this struct has been initially templated, so putting T here wouldn't work. I could perhaps add a type template to DecimalCastOperation, or maybe you have some ideas? I saw elsewhere in the code that 'using' was applied for this situation, but I figured typedef is just a simpler alternative for that.

Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1610]: e.g I could emulate this
// Cast String -> Numeric

```
template <typename T>
struct IntegerCastData {
	using Result = T;
	Result result;
	bool seen_decimal;
};
struct IntegerCastOperation {
	template <class T, bool NEGATIVE>
	static bool HandleDigit(T &state, uint8_t digit) {
		using result_t = typename T::Result;
		if (NEGATIVE) {
			if (state.result < (NumericLimits<result_t>::Minimum() + digit) / 10) {
				return false;
			}

Comment (Mytherin) [reply src/common/operator/cast_operators.cpp:1610]: You should be able to put `TYPE` here directly now, no?

Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1610]: Thijs helped me to take a look at this and suggested that I use decltype(state.result) instead, would that be ok?

Comment (LindsayWray) [reply src/common/operator/cast_operators.cpp:1610]: TYPE is set in a different scope, so I can't use it directly here

Comment (Mytherin) [reply src/common/operator/cast_operators.cpp:1610]: Ah apologies, yes you are correct. The github view hiding the code in the middle hid the scoping from me. The way it is set up now is fine.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**